### DISCUSSION TITLE
[RFC-004] Module system

### DISCUSSION THREAD

**Discussion by dpeter99** [Candle-Fire/umbra#5] [RFC]
My original plan was never fully realized. What is there is what I managed to scribble down in time for the thesis.
My original plan:
- There are assemblies. these are the DLLs or similar packages, a unit of code on the file system that can be reloaded.
- There are modules, a unit inside the core update loop, that can be ordered. 
So the idea would be that there is a class
- AssemblyManager (I think you did a similar one called Library Manager) that can load in new Assemblies and run them. This is what is responsible for also watching file system changes and detecting if one of them needs to be reloaded.
- Module Manager. This is what instantiates and orders the modules.

The things that are not well defined and are a stopgap currently:
- where the list of modules to load in comes from. Currently, the list of known modules is from each DLL's entry point that is called by the AssemblyManagger (would be if it was written, will be in the future). This is an interesting place cause it can add/remove modules based on some runtime check. But it also means we don't know what modules are in a DLL until it is linked and loaded. There were ideas about using JSON and a more comprehensive package setup. This would need more support from all parts like: 
  - Build a system to produce the correct package layout
  - Maybe bundle files? but bundling DLLs and loading them in from memory is a no-go.

- I'm not a fan of us not having compile-time constants for the Module IDs

- Ability to add tags to modules so you can easily say you depend on ALL renderers and not have to define an exact ID

# Hot Reload
So a hot reload would look like this:
1. The Assembly Manager gets notified by the OS that the file changed
2. The assembly manager tells the Module Mgr to unload all modules that are from that assembly
3. The module manager checks if the modules from there can be unloaded (not marked as "no-reload")
4. Tells all other modules that there is a reload happening (Probably 2 events: "Dependent Modules is reloading", "AnyModule Is reloading" for saving the world etc)
5. The modules are unloaded
6. reload
7. tell other Modules to reload

**Comment by TheCurle** [comment]
> AssemblyManager (I think you did a similar one called Library Manager)

`Library` is specifically for loading .lib or similar files, getting them into the system process, and pulling out function pointers.
It's a simple system-agnostic wrapper around dlopen() and dlsym().

It does not handle lifespans or reloading them.

**Comment by TheCurle** [comment]
> Ability to add tags to modules so you can easily say you depend on ALL renderers and not have to define an exact ID

If we do this, we need to associate a tag with an interface (a class with only pure virtual functions), so that we can reference the module's contents in a compile-time-disconnected way. Being able to say "i want any renderer" and then being forced to have a reference to, say, RxVkRenderer::instance would defeat the entire point.

The Module system should provide this interface for every tag supported by the engine.

This also means that there should be a finite, hardcoded, and immutable set of tags that we support.

**Comment by TheCurle** [comment]
> The Assembly Manager gets notified by the OS that the file changed
    The assembly manager tells the Module Mgr to unload all modules that are from that assembly
    The module manager checks if the modules from there can be unloaded (not marked as "no-reload")
    Tells all other modules that there is a reload happening (Probably 2 events: "Dependent Modules is reloading", "AnyModule Is reloading" for saving the world etc)
    The modules are unloaded
    reload
    tell other Modules to reload

This seems like a problem.

Let's reorganize this a little bit.
A lot of modules (for instance, modules that perform a similar function to the SDL module, who set up some global state and hold a single pointer upon which almost all the rest of the engine state depends) can simply migrate their state to the new assembly upon being asked.

We therefore have two stages: notify, reload

1. The Assembly Manager gets notified of a reload
2. The Assembly Manager loads the new assembly, as a copy, without running any static code
3. The Assembly Manager notifies the Module Manager
4. The Module Manager notifies affected modules of the impending migration, offering the new assembly.
5. The Module Manager checks whether modules of the new assembly are marked ready from that notification.
6. The Module Manager notifies the dependencies of any not-ready modules that a reload is about to happen.
7. The Module Manager notifies all modules that a reload is about to happen.
8. The Module Manager reloads all not-ready modules in the assembly being swapped, running static code at this time.
9. All system state is preserved, and the engine need not reload things unnecessarily, and therefore potentially risk corruputing global state.
